Updates to the project made by Dov

danimoh opened the pull request:
These changes were made by Dov and had been merged from https://github.com/DovAzencot/cashlink-generator into temporary branch `update-dov` and shall upon review be merged into branch `update` which serves as the development branch until the update is concluded.

```
Improvements:

- Complete TypeScript rewrite with enhanced type safety
- Modern environment configuration handling
- Updated dependencies for latest Node.js compatibility
- Nimiq 2.0 (Albatross PoS) support
- Improved development experience with proper typing
- Enhanced QR code generation capabilities
```
See changes to README included in this PR.

DovAzencot committed:
Cherry pick changes from github.com/DovAzencot/cashlink-generator

Cherry pick project update f8b73aa6 ("first commit") by DovAzencot from separate
remote git@github.com:DovAzencot/cashlink-generator.git, which is a fork of this
repository, which was not initialized as a fork though, i.e. it has an unrelated
commit history.
It could have been merged via --allow-unrelated-histories instead of a cherry-
pick, but a cherry-pick avoids the unrelated commit history to better preserve
git blame annotations.

danimoh commented on .env.example:
```diff
@@ -0,0 +1,14 @@
+TOKEN_LENGTH=6
+
+# Salt for cashlink generation (base64 encoded)
+SALT=your-base64-encoded-salt-here
```
Should add a note that this salt must be kept absolutely secret.

danimoh commented on LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2025 Dov Azencot
```
Assuming copyright on the whole thing is maybe a bit much.

DovAzencot replied:
I did mention you in the readme, obviously forgot to add you here, sorry! <3

danimoh commented on package.json:
```diff
@@ -1,15 +1,27 @@
 {
-  "name": "cashlink-generator",
+  "name": "cashlink",
```
This is not a Cashlink library, so I think `cashlink-generator` was the more suitable name.

danimoh commented on package.json:
```diff
@@ -1,15 +1,27 @@
+    "qr": "NODE_OPTIONS='--no-deprecation' tsx src/render-qr-codes.ts"
+  },
+  "keywords": [],
+  "author": "Dov Azencot",
```
This as well should stay `Nimiq, nimiq.com`.

danimoh commented on package.json:
```diff
@@ -1,15 +1,27 @@
+  "keywords": [],
+  "author": "Dov Azencot",
   "license": "MIT",
+  "packageManager": "pnpm@9.12.2+sha512.22721b3a11f81661ae1ec68ce1a7b879425a1ca5b991c975b074ac220b187ce56c708fe5db69f4c962c989452eee76c82877f4ee80f474cebd61ee13461b6228",
```
Haha, not a big fan of pnpm but alright, alright ;)
(Maybe I'll keept it. Maybe.)

danimoh commented on scripts/create-secret.ts:
```diff
@@ -1,21 +1,85 @@
+            resolve(answer);
+        });
+    });
+}
```
Might want to move this to a separate file to avoid duplication with `index.tx`.

danimoh commented on scripts/create-secret.ts:
```diff
@@ -1,21 +1,85 @@
+
+# Salt for cashlink generation (base64 encoded)
+SALT=${secretBase64}
+`;
```
Code style: Personally, I don't like template strings that break the code layout. I think as a team we also usually don't do that.

danimoh commented on scripts/create-secret.ts:
```diff
@@ -1,21 +1,85 @@
+# Length of cashlink tokens in characters
+TOKEN_LENGTH=6
+
+# Salt for cashlink generation (base64 encoded)
```
Also here a note should be added that the salt must be kept secret.

danimoh commented on scripts/create-secret.ts:
```diff
@@ -1,21 +1,85 @@
+# Salt for cashlink generation (base64 encoded)
+SALT=${secretBase64}
+`;
+        envContent = envContent || defaultEnv;
```
This does not properly update the `.env` file if a file existed previously with empty salt.

danimoh commented on scripts/create-secret.ts:
```diff
@@ -1,21 +1,85 @@
 
-createSecret();
+// Execute immediately
+createSecret().catch(error => {
```
```suggestion
createSecret().catch((error) => {
```
Code style: Usually, we put parenthesis around arrow function parameters.
(This project might have deviated because it's so old and has no linter on it.)

danimoh commented on scripts/create-secret.ts:
```diff
@@ -1,21 +1,85 @@
+createSecret().catch(error => {
+    console.error('Error:', error instanceof Error ? error.message : 'Unknown error');
+    process.exit(1);
+});
```
While generally thinking about error handling is a good idea, I don't think that catching errors here on the main method is beneficial. Unhandled exceptions terminate the node process anyway and provide more information like the stack trace.

danimoh commented on src/cashlink-statistics-handler.ts:
```diff
@@ -1,6 +1,36 @@
+import { Address } from '@nimiq/core';
+import { RpcClient } from './rpc-client';
+import { Cashlink } from './cashlink';
+import { Transaction as RpcTransaction } from '@blouflash/nimiq-rpc';
```
Can be type-only imports.

danimoh commented on src/cashlink-statistics-handler.ts:
```diff
@@ -9,94 +39,138 @@ async function createStatistics(cashlinks, reclaimAddress, timeZone, rpcClient)
         day: '2-digit',
     });
 
-    // Note that for reused cashlinks, these values might sum up to higher values than the cashlink count.
```
Nimiq Cashlinks do support being claimed multiple times, or fractional claiming, and also refunding.
This is still the case, so the comment should stay. Same for the one below.

danimoh commented on src/cashlink-statistics-handler.ts:
```diff
@@ -9,94 +39,138 @@ async function createStatistics(cashlinks, reclaimAddress, timeZone, rpcClient)
-    const claimsPerAddress = new Map(); // userfriendly address -> number
-    const claimsPerDate = new Map(); // date string in format yyyy-mm-dd -> [claims, claims only counting first timers]
+    const claimsPerAddress = new Map<string, number>();
+    const claimsPerDate = new Map<string, [number, number]>();
```
The comment on what the strings and numbers represent does still provide value.

danimoh commented on src/cashlink-statistics-handler.ts:
```diff
@@ -9,94 +39,138 @@ async function createStatistics(cashlinks, reclaimAddress, timeZone, rpcClient)
-                        previousDateClaimsFirstTimers + (!previousAddressClaims ? 1 : 0),
-                    ]);
+                for (const tx of transactionObjects) {
+                    const toAddress = tx.to; // Use recipient instead of toAddress
```
I don't understand what this comment refers to.

danimoh commented on src/cashlink-statistics-handler.ts:
```diff
@@ -9,94 +39,138 @@ async function createStatistics(cashlinks, reclaimAddress, timeZone, rpcClient)
-                    const previousAddressClaims = claimsPerAddress.get(toAddress) || 0;
-                    claimsPerAddress.set(toAddress, previousAddressClaims + 1);
-                    const [{value: month},,{value: day},,{value: year}] = dateFormatter.formatToParts(timestamp * 1000);
-                    const date = `${year}-${month}-${day}`; // lexically sortable
```
I do wonder, why did you remove such comments?

danimoh commented on src/cashlink-statistics-handler.ts:
```diff
@@ -9,94 +39,138 @@ async function createStatistics(cashlinks, reclaimAddress, timeZone, rpcClient)
         .filter(([, claims]) => claims > 1)
-        .sort(([, claimsA], [, claimsB]) => claimsB - claimsA); // sort descending
-    const claimsPerDateSorted = [...claimsPerDate]
-        .sort(([dateA], [dateB]) => (dateA < dateB ? -1 : dateA > dateB ? 1 : 0)); // sort ascending
```
Also here, why remove such comments?

danimoh commented on src/cashlink-statistics-handler.ts:
```diff
@@ -9,94 +39,138 @@ async function createStatistics(cashlinks, reclaimAddress, timeZone, rpcClient)
+        
+    const percentFormatter = new Intl.NumberFormat('en-US', { 
+        style: 'percent', 
+        maximumFractionDigits: 2 
```
Code style: We do usually include a trailing comma also on the last entry in object and array definitions and arguments lists etc.

danimoh commented on src/qr-code.ts:
```diff
@@ -1,4 +1,4 @@
-// Taken from https://github.com/nimiq/qr-encoder but with small changes and exported as node module
```
Why remove such attributions to the original source of the code and their authors?

DovAzencot replied:
Most of the comments regarding removal of comments were done by AI, I like to use it to creates docs and comments as it is usually very good a just that. I didn't notice that in the process it removed previously added comments, I'll make sure to check that next time.

danimoh commented on src/rpc-client.ts:
```diff
@@ -1,95 +1,263 @@
+     */
+    async isConsensusEstablished(): Promise<boolean> {
+        const response = await this._client.consensus.isConsensusEstablished();
+        return Boolean(response.data);
```
Could be slightly nicer with destructuring.
```suggestion
        const { data } = await this._client.consensus.isConsensusEstablished();
        return Boolean(data);
```

danimoh commented on src/rpc-client.ts:
```diff
@@ -1,95 +1,263 @@
+     * @returns Promise resolving to transaction details
+     * @throws If transaction not found or hash invalid
+     */
+    async getTransactionReceipt(txHash: string): Promise<Transaction> {
```
```suggestion
    async getTransaction(txHash: string): Promise<Transaction> {
```
The method should be renamed to `getTransaction` because you're now in fact requesting a full transaction. A transaction receipt on the other hand is a receipt for transaction inclusion into the chain which includes metadata like the transaction hash, block hash and block height, but no details about the transaction.

danimoh commented on src/rpc-client.ts:
```diff
@@ -1,95 +1,263 @@
+        const response = await this._client.blockchain.getTransactionByHash(txHash);
+        if (!response.data) {
+            throw new Error('Transaction not found');
+        }
```
Why are additional checks included here but not in other methods like `getBalance` for example.

danimoh commented on src/rpc-client.ts:
```diff
@@ -1,95 +1,263 @@
+        if (!response.data) {
+            throw new Error('Transaction not found');
+        }
+        return response.data as Transaction;
```
Why is the type cast necessary? `@blouflash/nimiq-rpc` is fully typed, isn't it?

danimoh commented on .DS_Store:
Should not be part of the code base.

danimoh commented on README.md:
There are some additional details in the old README which have been removed but would be good to preserve.

danimoh commented on src/rpc-client.ts:
```diff
@@ -1,95 +1,263 @@
+                BigInt(params.fee || 0),
+                blockHeight,
+                5
+            );
```
The transaction data in `params.data` is being ignored, which especially also means that Cashlinks are generated without the special Cashlink extra data.

danimoh commented on src/rpc-client.ts:
```diff
@@ -1,95 +1,263 @@
+            transaction.sign(params.sender);
+
+            const data = await this._client.call<any>({ method: 'sendRawTransaction', params: [ BufferUtils.toHex(transaction.serialize())] });
+            return data.data;
```
Nicer with destructuring. Similarly at other places.
```suggestion
            const { data } = await this._client.call<any>({ method: 'sendRawTransaction', params: [ BufferUtils.toHex(transaction.serialize())] });
            return data;
```

danimoh commented on src/rpc-client.ts:
```diff
@@ -1,95 +1,263 @@
+            // Sign the transaction
+            transaction.sign(params.sender);
+
+            const data = await this._client.call<any>({ method: 'sendRawTransaction', params: [ BufferUtils.toHex(transaction.serialize())] });
```
There's a method `sendRawTransaction` provided by `@blouflash/nimiq-rpc` which can be used here directly.

DovAzencot replied:
I tried it and didn't get it to work, that's why I used it like that, maybe I used it wrong, I didn't investigate further.

danimoh commented on src/rpc-client.ts:
```diff
@@ -1,95 +1,263 @@
+    /**
+     * Sends a raw transaction using direct transaction creation and signing
+     * @param params - Raw transaction parameters including KeyPair for signing
+     * @returns Promise resolving to transaction result
```
The method has no return type annotation and because the result of `call<any>` is being returned, the automatically determined return type will be `any`, which is not very elegant. I believe the actual return type is `Promise<string>` but I haven't confirmed and did not check what the string represents, probably a hex encoded transaction or something like that.

danimoh commented on src/rpc-client.ts:
```diff
@@ -1,95 +1,263 @@
+    /**
+     * Imports a wallet key
+     * @param keyData - The key data to import
+     * @returns Promise resolving to import result
```
This documentation is not much help. What is the import result?

danimoh commented on src/rpc-client.ts:
```diff
@@ -1,95 +1,263 @@
+
+            const response = await this._client.blockchain.getTransactionsByAddress(
+                address,
+                //@ts-ignore
```
`@ts-ignore` comments should have a description what they're ignoring.

danimoh commented on src/config.ts:
```diff
@@ -1,13 +1,70 @@
+ * @returns Validated string value
+ * @throws If environment variable is not set
+ */
+function assertEnvVar(value: string | undefined, name: string): string {
```
Can even consider turning this into a [typescript assertion function](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions), but would lead to more verbose code where it's used below.
```suggestion
function assertEnvVar(value: string | undefined, name: string): asserts value {
```

danimoh commented on src/config.ts:
```diff
@@ -1,13 +1,70 @@
+ * @param network - Network identifier ('main' or 'test')
+ * @returns Complete base URL for cashlinks
+ */
+export function getCashlinkBaseUrl(network: string): string {
```
`network` can be more tightly typed as `'main' | 'test'`.

onmax replied:
Soeren, if you read this and if you know if we have this type somewhere in `@nimiq/core`, I will buy you a beer ;)

I have created this type half a dozen times now. If it is not part of the lib, maybe I can create a PR to expose an enum or similar?

sisou replied:
You mean just a type that is the union of the two strings `"main" | "test"`? That doesn't exist in core, because there the network is called "MainAlbatross" (case-insensitive), but "main-albatross" is also accepted.

What I did last time I had this as an input was to check the string if it _contains_ "main" instead of strict equality.

onmax replied:
Yep, I know we have this https://github.com/nimiq/core-rs-albatross/blob/3e923ca18ea0ef1d2eaf275da8e291dbfacbd417/primitives/src/networks.rs#L73-L87

and there are multiple correct strings, just wondering if we had that in the TS types

sisou replied:
Ah. No, we don't have a `NetworkId` enum in @nimiq/core

danimoh commented on src/cashlink-statistics-handler.ts:
```diff
@@ -1,6 +1,36 @@
+import { Transaction as RpcTransaction } from '@blouflash/nimiq-rpc';
+
+/**
+ * Creates comprehensive statistics for a set of cashlinks
```
I think proper punctuation in documentation would be good (having sentences end with a dot).

danimoh commented on src/config.ts:
```diff
@@ -1,13 +1,70 @@
+        nodePort: assertEnvVar(process.env.NODE_PORT, 'NODE_PORT'),
+        network,
+        tokenLength: Number(assertEnvVar(process.env.TOKEN_LENGTH, 'TOKEN_LENGTH')),
+        salt: assertEnvVar(process.env.SALT, 'SALT'),
```
`assertEnvVar` can be turned into a `getEnvVar` which reads the value directly from `process.env[name]` instead of passing the value down and thereby remove the duplication of names in the call.

danimoh commented on src/cashlink.ts:
```diff
@@ -1,76 +1,154 @@
+    EASTER = 4,
+    GENERIC = 5,
+    BIRTHDAY = 6
+}
```
It's better to import the `CashlinkTheme` from `@nimiq/hub-api` as it was done previously, because new themes might be added there.

danimoh commented on src/cashlink.ts:
```diff
@@ -1,76 +1,154 @@
+ */
+export const CashlinkExtraData = {
+    FUNDING: new Uint8Array([0, 130, 128, 146, 135]),
+    CLAIMING: new Uint8Array([0, 139, 136, 141, 138]),
```
Also here, why have the documentation comments been removed?

danimoh commented on src/cashlink.ts:
```diff
@@ -1,76 +1,154 @@
+        const normalizedBaseUrl = baseUrl.endsWith('/') ? baseUrl.slice(0, -1) : baseUrl;
+        const finalBaseUrl = normalizedBaseUrl.includes('/cashlink') ? 
+            normalizedBaseUrl : 
+            `${normalizedBaseUrl}/cashlink`;
```
- Why is this necessary?
- The comment is wrong. This snippet does not only ensure the /cashlink/ is preserved, but also adds it if it's not present. Additionally, the trailing / is in fact not preserved or added, but removed.
- A better place for such a check would be the constructor, such that this invariant is also enforced for Cashlinks that are not created via the `parse` method.
- If the snippet is in fact needed (which I don't think), it can be simplified, under the assumption that the `/cashlink` part is supposed to be at the end of the URL.
```suggestion
        // Ensure the URL ends with /cashlink
        const normalizedBaseUrl = baseUrl.replace(/(?:\/cashlink)?\/?$/, '/cashlink');
```

danimoh commented on src/cashlink.ts:
```diff
@@ -1,76 +1,154 @@
+     * @returns New Cashlink instance
+     */
+    static parse(str: string): Cashlink {
+        const [baseUrl, rawHash] = str.split('#');
```
As you're also doing some post processing on the base URL, `rawBaseUrl` would be a name more consistent to `rawHash`. I do however believe that the post processing on the base URL is in fact not needed at all.

danimoh commented on src/cashlink.ts:
```diff
@@ -1,76 +1,154 @@
+    private _messageBytes: Uint8Array;
+    private _theme: number;
+    private static _textEncoder = new TextEncoder();
+    private static _textDecoder = new TextDecoder();
```
Code style: we typically list static properties before other properties, and static methods before other methods.

danimoh commented on src/cashlink.ts:
```diff
@@ -1,76 +1,154 @@
-        if (!NumberUtils.isUint8(messageBytes.byteLength)) throw new Error('Cashlink message is too long');
+    set message(message: string) {
+        const messageBytes = Cashlink._textEncoder.encode(message);
+        if (!NumberUtils.isUint8(messageBytes.byteLength)) {
+            throw new Error('Cashlink message is too long');
+        }
```
Code style: we _do_ typically enforce curly braces around the body of an `if` statement even if it has only one statement. However, exceptions are `return` and `throw` statements, which can be enclosed by curly braces but don't have to be. So the change here is fine, but according to our coding style wouldn't have been necessary.

danimoh commented on src/cashlink.ts:
```diff
@@ -79,7 +157,8 @@ class Cashlink {
         );
 
-        this._keyPair.privateKey.serialize(buf);
+        // Fix the serialization
```
What is meant by this comment?

danimoh commented on src/cashlink-transaction-handler.ts:
```diff
@@ -1,69 +1,93 @@
-module.exports = { fundCashlinks, claimCashlinks };
 
+    console.log(`Processed ${processed} Cashlinks, of which ${unclaimed} were unclaimed and redeemed now.`);
+}
```
Code style: usually we end our files with a trailing newline.

danimoh commented on src/file-handler.ts:
```diff
@@ -1,36 +1,122 @@
+    cashlinks: Map<string, Cashlink>;
+    shortLinks: Map<string, string>;
+    imageFiles: Map<string, string>;
 }
```
Would be good if the documentation included info on what the strings represent.

danimoh commented on src/file-handler.ts:
```diff
@@ -1,36 +1,122 @@
+    const lines = [...cashlinks].map(([token, cashlink]) => {
+        const shortLink = shortLinks?.get(token) || '';
+        const imageFile = imageFiles.get(token) || '';
+        const privateKeyBase64 = BufferUtils.toBase64Url(cashlink.keyPair.privateKey.serialize());
```
In the previous implementation, private keys were base64 encoded, not base64url, so this change is not backwards compatible and there is no good reason to change it.

danimoh commented on src/file-handler.ts:
```diff
@@ -1,36 +1,122 @@
+        const url = cashlink.render();
+        const urlObj = new URL(url);
+        const baseWithCashlink = urlObj.origin + (urlObj.pathname.includes('/cashlink/') ? urlObj.pathname : '/cashlink/');
+        const finalUrl = `${baseWithCashlink}#${urlObj.hash.substring(1)}`;
```
- Why do you want to ensure that `/cashlink/` is included?
- Instead of enforcing this on export, it would be better if the `Cashlink` class would enforce it itself, which it currently does only on `parse`.
- The original URL path is entirely discarded, if it doesn't include `/cashlink/`. That's not desirable.
- I don't think this is actually useful.

danimoh commented on src/file-handler.ts:
```diff
@@ -1,36 +1,122 @@
+            // Parse the cashlink URL to get parameters
+            const url = new URL(cashlinkUrl);
+            const hashPart = url.hash.substring(1); // Remove leading #
+            const [encodedData] = hashPart.split(',');
```
There's no `,` expected to be in the encoded cashlink.

danimoh commented on src/file-handler.ts:
```diff
@@ -1,36 +1,122 @@
+            const decodedData = BufferUtils.fromBase64Url(encodedData);
+            
+            // Create a proper SerialBuffer to correctly read the uint64 value
+            const buf = new SerialBuffer(decodedData);
```
`decodedData` already is a `SerialBuffer`.

danimoh commented on src/file-handler.ts:
```diff
@@ -1,36 +1,122 @@
+
+            // Create cashlink with parsed data
+            const cashlink = new Cashlink(
+                url.origin + (url.pathname.includes('/cashlink') ? url.pathname : '/cashlink'),
```
See review comment on lines 112 to 116.

danimoh commented on src/file-handler.ts:
```diff
@@ -1,36 +1,122 @@
+                url.origin + (url.pathname.includes('/cashlink') ? url.pathname : '/cashlink'),
+                keyPair,
+                value,
+                message
```
Missing parsing and passing `theme`. This means that any custom theme is discarded on import.

danimoh commented on src/file-handler.ts:
```diff
@@ -1,36 +1,122 @@
+                keyPair,
+                value,
+                message
+            );
```
No need to duplicate parsing logic. There's `Cashlink.parse` which takes care of exactly that and is also what in fact was used before.

danimoh commented on src/file-handler.ts:
```diff
@@ -1,36 +1,122 @@
         }
-        imageFiles.set(token, imageFile);
     }
-    if (cashlinks.size === 0) throw new Error('No cashlinks imported.');
```
This check still makes sense in my opinion.

danimoh commented on src/render-qr-codes.ts:
```diff
@@ -1,21 +1,71 @@
+ * The generator supports both programmatic usage and CLI operation.
+ */
 
+import { createCanvas, Canvas, CanvasRenderingContext2D } from 'canvas';
```
`Canvas` and `CanvasRenderingContext2D` are used as types only.
```suggestion
import { createCanvas, type Canvas, type CanvasRenderingContext2D } from 'canvas';
```

(also in `render-coins.ts`)

danimoh commented on src/render-qr-codes.ts:
```diff
@@ -1,21 +1,71 @@
+    const context = canvas.getContext('2d');
     context.quality = 'best';
-    context.patternQuality = 'best'; // Affects pattern (gradient, image, etc.) rendering quality.
+    context.patternQuality = 'best';
```
Comment still useful.
Also in `render-qr-codes.ts`.

danimoh commented on src/render-qr-codes.ts:
```diff
@@ -38,43 +95,50 @@ function renderQrCodes(links, folder, options = {}) {
-if (require.main === module) {
-    // We're run as `node render-qr-codes.js`.
-    // Provide a little utility for rendering a single qr code in Nimiq style.
+export default renderQrCodes;
```
The method can be directly exported above where it's defined.

danimoh commented on src/render-qr-codes.ts:
```diff
@@ -38,43 +95,50 @@ function renderQrCodes(links, folder, options = {}) {
         createQrCode(filepath, content, {
             fill: {
                 type: 'radial-gradient',
-                // circle centered in bottom right corner with radius of the size of qr code diagonal
```
Another comment to be added back.

danimoh commented on src/render-coins.ts:
```diff
@@ -21,29 +36,63 @@ const BACK_LOGO_SIZE = calculateHexHeight(HEXAGON_RADIUS) * (1 - 1 / 1.618); //
     // Note that the radius is the same as the full side lengths.
     const height = calculateHexHeight(radius);
 
-    // Corners of the hexagon if it wouldn't be rounded.
```
And one more useful comment to preserve.

danimoh commented on src/render-coins.ts:
```diff
@@ -109,15 +158,16 @@ function drawHexagon(centerX, centerY, radius, borderRadius, context) {
+        ] as [number, number];
+        
+        const startAngle = arcPointToAngle(lines[i].end as [number, number], center, borderRadius);
+        const endAngle = arcPointToAngle(lines[j].start as [number, number], center, borderRadius);
```
Are the type assertions needed?

danimoh commented on src/render-coins.ts:
```diff
@@ -242,13 +320,27 @@ function drawBack(centerX, centerY, size, context) {
+ * @param side - Which sides to render ('both', 'front', or 'back')
+ * @returns Map of token to generated image filenames
+ */
+export function renderCoins(
```
It was the default export before, so should probably still be, also for consistency with `render-qr-codes.ts`, where the render method is the default export.

danimoh commented on src/index.ts:
```diff
@@ -1,44 +1,53 @@
+import readline from 'readline';
+import { Writable } from 'stream';
+import { MnemonicUtils } from '@nimiq/core';
+import { BufferUtils, SerialBuffer, Hash, PrivateKey, KeyPair, PublicKey, Address } from '@nimiq/core';
```
These imports can be merged.

danimoh commented on src/index.ts:
```diff
@@ -1,44 +1,53 @@
+import renderQrCodes from './render-qr-codes';
+import { claimCashlinks, fundCashlinks } from './cashlink-transaction-handler';
+import { createStatistics } from './cashlink-statistics-handler';
+import path from 'path';
```
Code style: usually we order imports of dependencies before imports of local files.

danimoh commented on src/index.ts:
```diff
@@ -47,256 +56,297 @@ function getCurrentDateString() {
-        + '/0..255' // ability to specify theme as a number for themes that are not defined in HubApi yet
+        + '/0..255'
         + (oldCashlinkTheme !== undefined
-            ? `; old theme: ${CashlinkTheme[oldCashlinkTheme].toLowerCase() || oldCashlinkTheme}` // reverse map or num
```
I don't see why such explanatory commens should be removed.

danimoh commented on src/index.ts:
```diff
@@ -47,256 +56,297 @@ function getCurrentDateString() {
+ * Securely prompts for private key input
+ * Hides input from display and processes backup words
+ * @returns Promise resolving to private key bytes
+ */
```
The info of the removed comment should at least be included in the documentation…

danimoh commented on src/index.ts:
```diff
@@ -47,256 +56,297 @@ function getCurrentDateString() {
-
-    const backupWords = [];
-    rl.on('line', (line) => {
-        // split at whitespace and strip numbers for direct copying from Nimiq Keyguard
```
Why remove this? Why??

danimoh commented on src/index.ts:
```diff
@@ -47,256 +56,297 @@ function getCurrentDateString() {
-    });
+    }) as MutableStdout;
+
+    mutableStdout.muted = false;
```
This value will be set again below, so it doesn't need to be set here.

danimoh commented on src/index.ts:
```diff
@@ -47,256 +56,297 @@ function getCurrentDateString() {
 
-function createCashlinks(cashlinkCount, cashlinkValue, cashlinkMessage, cashlinkTheme) {
-    const cashlinks = new Map(); // token -> cashlink
-    // secret salt to deterministically calculate cashlinks from random tokens
```
Another comment to preserve.

danimoh commented on src/index.ts:
```diff
@@ -47,256 +56,297 @@ function getCurrentDateString() {
+    STATISTICS: 'statistics',
+} as const;
+
+type OperationType = typeof Operation[keyof typeof Operation];
```
Would be better as an `enum`.

danimoh commented on src/index.ts:
```diff
@@ -47,256 +56,297 @@ function getCurrentDateString() {
         const links = shortLinks
             || new Map([...cashlinks].map(([token, cashlink]) => [token, cashlink.render()]));
-        imageFiles = renderQrCodes(links, folder);
+        imageFiles = renderQrCodes(links, folder); // Remove Object.entries - it's already a Map
```
Don't know what this comment refers to.

danimoh commented on src/index.ts:
```diff
@@ -47,256 +56,297 @@ function getCurrentDateString() {
     console.log('\nBefore funding the Cashlinks, please check the generated assets.');
     console.log('To continue with funding, please import an account via its backup words to use for funding and make '
-        + `sure it holds at least ${requiredBalance / 1e5} NIM (of which ${totalFees / 1e5} NIM fees).\n`
-        + 'Note that it\'s recommendable to create a new key only for this operation.');
```
The past showed that it's recommendable to create separate keys for for each Cashlink campaign for better bookkeeping, for example if unclaimed Cashlinks are to be reclaimed, they can be reclaimed to the funding address that was created just for that campaign and then be sent back to some more generic marketing account from there. Using a regularly used key instead of a separate key is not recommended and can be inconvenient because Cashlink generation and reclaiming leads to a lot of transactions on that address which clutter the transaction history and are extra effort to sync in the Wallet because it matches each Cashlink with its final recipient, so has to end up querying each Cashlink address as well.

danimoh commented on src/index.ts:
```diff
@@ -47,256 +56,297 @@ function getCurrentDateString() {
+        rl.on('close', () => {
+            const extendedPrivateKey = MnemonicUtils.mnemonicToExtendedPrivateKey(backupWords.join(' '));
+            const privateKey = extendedPrivateKey.derivePath(`m/44'/242'/0'/0'`).privateKey;
+            resolve(privateKey.serialize()); 
```
Would be better to return the `PrivateKey` instance instead of the bytes, as it was the case before, because otherwise the first thing that needs to be done after calling `promptPrivateKey` in your code is to deserialize it again.

danimoh commented on src/index.ts:
```diff
@@ -47,256 +56,297 @@ function getCurrentDateString() {
+    const isUnlocked = await rpcClient.unlockWalletAccount(userFriendlyAddress);
+    if (!isUnlocked) {
+        throw new Error('Failed to unlock account');
     }
```
I preferred the previous approach where the key was only temporary held in memory and not imported, which was better suited to using a separate account for each campaign, without leaving behind a bunch of imported keys. Also, importing the key doesn't make all that much sense, if the user still has to provide the full recovery words / private key on usage anyway, as opposed to just providing the funding address of a key, which then only asks for the recovery words, if it's not imported yet.

danimoh commented on src/index.ts:
```diff
@@ -313,67 +363,215 @@ async function wizardCreateStatistics(cashlinks, folder) {
+        } catch (e) {
+            throw new Error(`Failed to parse cashlinks file: ${e instanceof Error ? e.message : 'Unknown error'}`);
+        }
+    } catch (error) {
```
The nested `try ... catch` doesn't provide much value.

danimoh commented on src/index.ts:
```diff
@@ -313,67 +363,215 @@ async function wizardCreateStatistics(cashlinks, folder) {
+        } else {
+            const createResult = await wizardCreateCashlinks();
+            cashlinks = createResult.cashlinks;
+            shortLinks = createResult.shortLinks;
```
Destructuring was used before and kept above, so why change it here?

danimoh commented on src/index.ts:
```diff
@@ -313,67 +363,215 @@ async function wizardCreateStatistics(cashlinks, folder) {
+            const oldImageFiles = imageFiles;
+            imageFiles = await wizardCreateImages(cashlinks, shortLinks, folder);
+            shouldExport = shouldExport || !oldImageFiles.size
+                || [...oldImageFiles.values()][0] !== [...imageFiles.values()][0];
```
The previous
```suggestion
                || oldImageFiles.values().next().value !== imageFiles.values().next().value;
```
was more efficient as it didn't entirely iterate over boths maps.

danimoh commented on src/index.ts:
```diff
@@ -313,67 +363,215 @@ async function wizardCreateStatistics(cashlinks, folder) {
+
+        if (shouldExport) {
+            console.log('Exporting cashlinks.');
+            const file = `${folder || '.'}/cashlinks${importResult ? ` (update ${getCurrentDateString()} ${operations.join(' ')})` : ''}.csv`;
```
The previous
```suggestion
            const file = `${folder || '.'}/cashlinks`
                + (importResult ? ` (update ${getCurrentDateString()} ${operations.join(' ')})` : '')
                + '.csv';
```
was more readable.

danimoh commented on src/index.ts:
```diff
@@ -313,67 +363,215 @@ async function wizardCreateStatistics(cashlinks, folder) {
     }
 
-    if (operations.includes(Operation.FUND)) {
-        // fund after export, to make sure the cashlinks were saved, if needed
```
Another comment to keep.

danimoh commented on src/index.ts:
```diff
@@ -313,67 +363,215 @@ async function wizardCreateStatistics(cashlinks, folder) {
+    catch (error) {
+        console.error('Error:', error instanceof Error ? error.message : 'Unknown error');
+        process.exit(1);
     }
```
While generally thinking about error handling is a good idea, I don't think that the `try ... catch` here is beneficial. Unhandled exceptions terminate the node process anyway and provide more information like the stack trace.
The same is arguably true for several of the other introduced `try ... catch` blocks, which additionally lead to a lot of indentation changes.

danimoh commented on src/index.ts:
```diff
@@ -313,67 +363,215 @@ async function wizardCreateStatistics(cashlinks, folder) {
+main().catch(error => {
+    console.error('Unhandled error:', error);
+    process.exit(1);
+});
```
The `main` method does not throw any errors, as it handles them itself. But even if it would, arguably it is better to also not handle them here, see above.

danimoh commented on test/index.ts:
```diff
@@ -0,0 +1,80 @@
+         */
+        const txHash = 'd15bd926ad0c0e52346badf29cc128a26369d13a45498ad5e6a122f986054132';
+        const transaction = await client.getTransactionReceipt(txHash);
+        console.log(`Transaction ${txHash} was sent on block: `,transaction.blockNumber);
```
`getTransactionReceipt` is actually not used by the project, so it can be removed from the tests and `rpc-client`. Instead, `getTransactionsByAddress` should be tested.

danimoh commented on src/index.ts:
```diff
@@ -1,44 +1,53 @@
+ * - Generate QR codes and coin images
+ * - Create usage statistics
+ * 
+ * The tool supports both interactive CLI usage and programmatic integration.
```
Programmatic integration (reading from command line parameters) is not implemented.

danimoh commented on src/cashlink-statistics-handler.ts:
```diff
@@ -9,94 +39,138 @@ async function createStatistics(cashlinks, reclaimAddress, timeZone, rpcClient)
-            let wasFunded = false;
-            let wasUserClaimed = false;
-            let wasReclaimed = false;
+        console.log(cashlinkUserFriendlyAddress)
```
This spams the logs quite a bit.

danimoh reviewed (commented):
Hi @DovAzencot,

thanks for the extensive update. The migration to Nimiq PoS seems to work nicely and I welcome the addition of documentation and the modernization of code to typescript and es6 modules. Overall and on a high level, I like this update. However, on close review there are of course also details I like less and which I want to give further feedback on:

- It would have been nice to submit this as a PR to the existing project, instead of as a separate project. That would have simplified reviewing against the existing code, leaving feedback and eventually merging the changes into the main project. I now cherry-picked your changes onto this project and opened this pull request to achieve basically that.
- This PR / update includes too much. It would have been better to first focus on the task at hand and then take it from there. That would also have simplified the review.
- The whole thing is a single commit. Better would have been well encapsulated commits per change or feature with well written commit messages.
- Assuming the copyright of the entire project is not cool.
- Apart from the good changes, there are also unnecessary, unrelated or bogus changes, and not all of those are even improvements. Some of those changes introduce bugs, add code that provides no benefit, change the coding style deviating from what our usual coding style is, or are bogus changes like changing the order of methods in a file, which further complicates diffing and reviewing.
- Many useful code comments have been removed for no obvious reason. Those are comments that the people that originally wrote the code put some thought into and added for a reason. I know, because for this project that's basically me :D

I did a full, detailed review of this PR. Further feedback and issues on relevant code snippets are given in the comments below. Please take a look at those.

You can continue working on this, if you want, but it's not expected. I will probably take this on at some point, but for now finalizing this PR is not the highest priority, and the functionality is basically already there for if Cashlinks need to be generated now.

danimoh committed:
Remove trailing whitespaces; ensure trailing new line at end of file

danimoh committed:
Change copyright back to Nimiq

danimoh committed:
Remove .DS_Store file

danimoh committed:
Add comments back that were blindly removed and remove AI prompts left behind

danimoh committed:
Many many code style changes and small fixes or code improvements

danimoh committed:
Type fixes and improvements

danimoh committed:
index.ts, create-secret.ts: Let exceptions terminate the process instead of catching them

Catching errors on the main methods / entry points isn't beneficial. Unhandled exceptions terminate
the node process anyway and provide more information like the stack trace.

Additionally, there was additional error handling at the invocation of the main methods, which is
entirely unnecessary, as the main methods already caught all exceptions themselves.

Also removed an unnecessary nested try..catch.

See
https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1938060249
https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1938063931
https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1937998026

danimoh committed:
rpc-client: Remove unused method getTransactionReceipt

- The method was also named incorrectly, as it in fact fetched a full transaction, not just a
  receipt.
- Also remove it from tests. Instead, add a test for getTransactionsByAddress.

See
https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1933569716
https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1938451266

danimoh committed:
rpc-client: use sendRawTransaction directly from @blouflash/nimiq-rpc instead of reimplementing

See https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1934013080

danimoh committed:
rpc-client: improve error handling

- Don't catch and rethrow errors, to preserve original stack traces.
- More consistent and correct checking for expected response.

See https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1933581832

danimoh committed:
rpc-client: properly handle tx data and network in sendRawTransaction

Transaction data was previously ignored, and the network was hardcoded as testnet.

See https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1933959582

danimoh committed:
cashlink,file-handler: remove unnecessary and broken enforcement of /cashlink in cashlink URL

The removed code attempted to preserve /cashlink in the URL (for what ever reason), but
in fact did not only preserve it, but also enforced it to be added. In file-handler, the
original url path would even be discarded entirely.
It was also not very elegant to begin with. A better place for handling of an invariant
like the url enforcing a specific format, would have been the Cashlink constructor instead
of doing and repeating it on parsing and exporting, but then also omitting it on Cashlinks
that are created via the constructor.

See https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1934545356
and https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1935980329

danimoh committed:
file-handler: use Cashlink.parse to import cashlinks, instead of broken duplicate implementation

The now removed duplicate implementation was missing handling of themes, entirely
discarding them on import and missing stripping of ~ separators, and also doing
some non-sensical things.

At the same time fix Cashlink.parse which did not seek the SerialBuffer correctly
forward after reading the private key since the migration to Nimiq PoS's @nimiq/core.

See
https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1933959582
https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1936143474
https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1936077341

danimoh committed:
file-handler: change export of private keys back to base64

That's what it previously was, and the change to base64Url was an unnecessary
breaking change.

See https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1935936912

danimoh committed:
index.ts: return privateKey instance instead of bytes in promptPrivateKey

This avoids extra serialization to bytes and immediate deserialization again where the
private key is used.

See https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1937294377

danimoh committed:
Keep user private key only temporary in memory, don't import them

Revert private key handling to the previous approach where the key was only temporary held
in memory and not imported, which didn't expose the key and was better suited to using a
separate account for each campaign, without leaving behind a bunch of imported keys. Also,
importing the key did't make all that much sense, if the user still has to provide the full
recovery words / private key on usage anyway, as opposed to just providing the funding address
of a key, which then only asks for the recovery words, if it's not imported yet.

See https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1937968924

danimoh committed:
index.ts: define Operations as enum

See https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1937100433

danimoh committed:
cashlink.ts: import CashlinkTheme from @nimiq/hub-api

This is to keep the Cashlink themes in sync with what the Hub actually defines.

See https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1934468287

danimoh committed:
config: simplify reading of env vars and improve sanity checks

See https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1934407560

danimoh committed:
create-secret: fix handling of empty secret in existing env files

- Previously, if a file existed but with empty secret, the script did not set
  the secret.
- If a file with empty secret existed, the script previously still asked wheter
  to overwrite the secret.

See https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1932861255

danimoh committed:
index.ts,create-secret,render-qr-codes: move duplicate prompt function to common utils

See https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1932852932

danimoh committed:
Readme: reintroduce the removed content and update some

See https://github.com/nimiq/cashlink-generator/pull/1#discussion_r1933707402

danimoh committed:
Switch to NodeNext style imports

danimoh committed:
Switch to yarn PnP

danimoh committed:
package.json: run scripts natively via node without tsx

danimoh commented:
All resolved comments have been addressed. I.e. everything but punctuation of the documentation. Ready to be merged :tada:

danimoh commented:
Manually merged to master, but kept the update branch as base branch here for better diffing.
